### PR TITLE
BUG: Convert TNC C module to cython

### LINKED ISSUE #14879: BUG: TNC output is different if a callback is used.
### Describe your issue.

The output from a TNC minimisation is different, depending on whether a callback is used or not. All a callback should do is call an external function with a *copy* of x, and it shouldn't modify the the course of the minimisation. The output should be identical, including `res.nfev, res.nit, res.x`. 

### Reproducing Code Example

```python
import numpy as np
from scipy.optimize import rosen, minimize

bounds=[(0, 2), (0.0, 4.0), (0.0, 10.0), (0.0, 10.0)]
rng = np.random.default_rng(1)
x0 = rng.random(size=4)

def callback(x):
    pass

res = minimize(rosen, x0, bounds=bounds, method='TNC')
res2 = minimize(rosen, x0, bounds=bounds, method='TNC', callback=callback)
np.testing.assert_allclose(res2.x, res.x)
np.testing.assert_allclose(res2.nfev, res.nfev)
```

### Error message

```shell
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_57329/376828189.py in <module>
----> 1 np.testing.assert_allclose(res2.x, res.x)

    [... skipping hidden 1 frame]

~/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    840                                 verbose=verbose, header=header,
    841                                 names=('x', 'y'), precision=precision)
--> 842             raise AssertionError(msg)
    843     except ValueError:
    844         import traceback

AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 4 / 4 (100%)
Max absolute difference: 2.49955172e-05
Max relative difference: 2.49957461e-05
 x: array([0.999998, 0.999992, 0.999983, 0.999966])
 y: array([0.999999, 0.999998, 0.999995, 0.999991])
```

### SciPy/NumPy/Python version information

'1.8.0.dev0+1830.eee6248'

### PR DESCRIPTION
**Pull request by andyfaff**
The TNC C module (`optimize.tnc.moduleTNC.c`) is quite hard to update/follow. This PR converts it to use a Cython wrapper instead.

This conversion will help debug #14879 (because I'll become familiar with the codebase)

### CODE REVIEW

**Comment by andyfaff** [comment]
@rgommers, when I try to build this cython extension (using `python setup.py build`) I get the following error message:

```
building extension "scipy.optimize._tnc_module" sources
Traceback (most recent call last):
  File "setup.py", line 622, in <module>
    setup_package()
  File "setup.py", line 618, in setup_package
    setup(**metadata)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/core.py", line 169, in setup
    return old_setup(**new_attr)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build.py", line 61, in run
    old_build.run(self)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build_src.py", line 144, in run
    self.build_sources()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build_src.py", line 161, in build_sources
    self.build_extension_sources(ext)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build_src.py", line 324, in build_extension_sources
    sources, py_files = self.filter_py_files(sources)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build_src.py", line 391, in filter_py_files
    return self.filter_files(sources, ['.py'])
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/numpy/distutils/command/build_src.py", line 400, in filter_files
    (base, ext) = os.path.splitext(source)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/posixpath.py", line 118, in splitext
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not list
```

Can you see anything incorrect with my alterations to `optimize.setup.py`?

**Comment by andyfaff** [comment]
@ev-br, there are a few things that need to be fixed further on, but your suggestion doesn't allow to go any further.

**Comment by andyfaff** [comment]
This PR is ready for a review now. The test fails are unrelated.

**Comment by andyfaff** [comment]
The privatisation drive meant that this PR was suffering from merge conflicts. It was easier to rebase on master and force push, after a bit of manual twiddling.

@da-woods, would it be possible for you to review the cython code for me? Hopefully there's nothing too egregious.
I've tried to copy the outgoing `moduleTNC.c` in terms of structure.

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    try:
+        xcopy = np.empty(n, dtype=float)
+        x_data = <float64_t *>np.PyArray_DATA(xcopy)
+        memcpy(x_data, x, sizeof(double) * n)
+    except MemoryError as exc:
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:109]: Are there any other errors possible here (for example, someone might succeed in monkeypatching `np` in `_tnc_module`)? I'd probably make this `BaseException` instead of `MemoryError`

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        # couldn't allocate memory
+        py_state.failed = 1
+        raise exc
+
+    val = (<object>py_state.py_callback)(xcopy)
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:114]: This can also raise an exception I imagine - should that set `py_state.failed = 1` too?

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        offset_data = <float64_t *>np.PyArray_DATA(offset)
+
+    if (n1 != n2 or
+        n != n1 or
+        (scale_data != NULL and n != n3) or
+        (offset_data != NULL and n != n4)):
+        raise ValueError("tnc: vector sizes must be equal")
+
+    x = np.copy(x0, order="C")
+    g = np.zeros_like(x, dtype=float)
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:191]: Should dtype be `np.float64` just to make it match the pointer case explicitly? - I don't know how strictly specified the size of `float` is (both now and in the future)

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    try:
+        xcopy = np.empty(n, dtype=float)
+        x_data = <float64_t *>np.PyArray_DATA(xcopy)
+        memcpy(x_data, x, sizeof(double) * n)
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:108]: This may not be necessary - it looks from tnc.c like `callback_function` receives a temporary copy of the data anyway.

But I could be misunderstanding.

**Review by da-woods — COMMENTED** [review]
There's nothing I can see that's obviously wrong. I've made a few small comments to prove I have read it.

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx:109]: I should probably make `np.empty(n, dtype=np.float64)` if I'm going to be casting to a float64 pointer.

If all I'm doing is re-raising the exception I guess it doesn't matter what the Exception type is I guess.

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 4 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx:108]: `tnc.c` has to use a temporary array because the in-place unscaling/scaling is altering `x` by a small amount. So it's necessary to memcpy to the temp array, unscale in-place, then callback with that temp array.

The comment here may be a little misleading. Here we need to call a Python function with the contents of `x`. Is there a more direct way of doing that than creating a numpy array, transferring the information in `x` to that array, then calling the Python callback function?

Comment (da-woods) [reply scipy/optimize/tnc/_tnc_module.pyx:108]: If you don't need to make a copy, then potentially you could use Cython memoryviews:

```
xview = np.asarray(<double[:n]>(x))  # probably don't use this code - see below!
```

That creates a Cython memoryview of `x` and then creates a Numpy array that wraps the view.

**However**, neither the Numpy array or the Cython memoryview actually own the memory. Therefore, if a user decides to store that array inside the callback (to plot it later, for example) then it will be invalid by the time they want it. 

I think your "create a Numpy array and copy `x` into it is probably the right thing to do.

Comment (da-woods) [reply scipy/optimize/tnc/_tnc_module.pyx:108]: Just to summarise, the upshot of all of that is that I think you should keep making the copy (like you are doing now)

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx:114]: At this point the callstack is:
```
Python --> Cython ([L197](https://github.com/scipy/scipy/pull/14882/files#diff-f29933549abd450f6377e8daef0e671508e127b1b92db3e6c9c3b376bda0c2c6R197)) --> C ([tnc.c L804](https://github.com/scipy/scipy/blob/master/scipy/optimize/tnc/tnc.c#L804)) --> Cython (L114) --> Python (callback)
```

I don't know how Cython works here. If the Python callback raises an exception, then how does the Cython code here propagate the exception? Specifically, will the memory cleanup at the end of the `tnc` function ([tnc.c, L909](https://github.com/scipy/scipy/blob/master/scipy/optimize/tnc/tnc.c#L909)) ever run? If it doesn't then would there be a memory leak?

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx:109]: I think the only reason to catch Exceptions here is to make sure that `py_state.failed` is always set to 1.

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    if not gx.flags['C_CONTIGUOUS']:
+        gx = np.ascontiguousarray(gx, dtype=np.float64)
+
+    g_data = <float64_t *>np.PyArray_DATA(gx)
+    memcpy(g, g_data, n * sizeof(double))
+
+    return 0
+
+
+cdef void callback_function(double x[], void *state) except *:
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:94]: There's an issue here that I didn't pick up on earlier (and probably should have done - sorry!).

Looking at the `tnc` code - there's no way for this function to indicate that it's failed (and therefore it can't stop the overall operation). This will potentially cause problems later when some other bit of Python code runs and finds that an exception is already set.

I think the solution is to add (at the start of `function`):

```
if py_state.failed:
    return 1
```

i.e. to have `function` check to make sure that `callback_function` hasn't failed on the previous call.

It's a little dodgy because it isn't handling the exceptions how quite Cython expects, but I think it should work providing it happens before anything Python-like is executed. I'm slightly surprised that the exception test currently passes.

File: scipy/optimize/tnc/_tnc_module.pyx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    # However, if the user function raises Exceptions Cython should deal
+    # with it.
+
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+        double *g_data
+
+    py_state = <pytnc_state *>state
```
Comment (da-woods) [inline scipy/optimize/tnc/_tnc_module.pyx:59]: So I'm proposing the `py_state.failed` check should go here.

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 5 above)
Comment (da-woods) [reply scipy/optimize/tnc/_tnc_module.pyx:94]: This also breaks with the current implementation in scipy:

```
>>> from scipy import optimize
>>> def callback(*a):
...   raise RuntimeError("xxx")
... 
>>> optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)
Traceback (most recent call last):
  File "<stdin>", line 2, in callback
RuntimeError: xxx

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/_minimize.py", line 622, in minimize
    return _minimize_tnc(fun, x0, args, jac, bounds, callback=callback,
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/tnc.py", line 414, in _minimize_tnc
    rc, nf, nit, x = moduleTNC.minimize(func_and_grad, x0, low, up, scale,
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 258, in fun_and_grad
    if not np.array_equal(x, self.x):
  File "<__array_function__ internals>", line 4, in array_equal
SystemError: PyEval_EvalFrameEx returned a result with an error set
```

The other option would just be to have `callback` swallow the exception (maybe printing it first). To be honest, that might be cleaner and simpler

**Comment by andyfaff** [comment]
>  The other option would just be to have callback swallow the exception (maybe printing it first). To be honest, that might be cleaner and simpler

Yet another way would be to modify the C code to have `tnc_callback` produce a non-void return.

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 6 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx]: I added that.

File: scipy/optimize/tnc/_tnc_module.pyx (hunk 5 above)
Comment (andyfaff) [reply scipy/optimize/tnc/_tnc_module.pyx]: When I run your example (with the PR in its current state) I get:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_19744/2729457978.py in <module>
      3   raise RuntimeError("xxx")
      4 
----> 5 optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)

~/Documents/Andy/programming/scipy/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    665                                callback=callback, **options)
    666     elif meth == 'tnc':
--> 667         res = _minimize_tnc(fun, x0, args, jac, bounds, callback=callback,
    668                             **options)
    669     elif meth == 'cobyla':

~/Documents/Andy/programming/scipy/scipy/optimize/_tnc.py in _minimize_tnc(fun, x0, args, jac, bounds, eps, scale, offset, mesg_num, maxCGit, maxiter, eta, stepmx, accuracy, minfev, ftol, xtol, gtol, rescale, disp, callback, finite_diff_rel_step, maxfun, **unknown_options)
    412             maxfun = max(100, 10*len(x0))
    413 
--> 414     rc, nf, nit, x, funv, jacv = moduleTNC.tnc_minimize(
    415         func_and_grad, x0, low, up, scale,
    416         offset, messages, maxCGit, maxfun,

moduleTNC.pyx in moduleTNC.tnc_minimize()

moduleTNC.pyx in moduleTNC.callback_function()

moduleTNC.pyx in moduleTNC.callback_function()

/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_19744/2729457978.py in callback(*a)
      1 from scipy import optimize
      2 def callback(*a):
----> 3   raise RuntimeError("xxx")
      4 
      5 optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)

RuntimeError: xxx
```

so it seems like there's no issue with the Exception raising at least. I attached the Python interpreter process to the macOS Instruments program and I can't detect any leaks when the callback exception is raised.

**Review by rgommers — CHANGES_REQUESTED** [review]
Let me add a review comment to remember to undo the test skips in gh-15053 once this PR is ready. In https://github.com/scipy/scipy/issues/15051#issuecomment-971401621 @andyfaff commented that this PR is likely to fix the precision issues in the now-skipped tests.

**Comment by andyfaff** [comment]
@rgommers, the PR is ready to go as far as I'm concerned. @da-woods had a look at the cython code and seemed to think it's in an ok state. I checked that raised exceptions works from both the objective and callback (with tests) and I can't detect any leaks.

**Comment by andyfaff** [comment]
@ev-br @rgommers it'd be great if this PR could go in before 1.8 is released (it closes . As far as I'm concerned it's ready to be merged. @da-woods has reviewed it, but it needs a core-dev to have a look at it and merge.

**Review by tylerjereddy — APPROVED** [review]
I was initially worried that the error handling after this change wasn't fully resolved in the discussion comments, but I see that:

1) some of those comments were marked outdated
2) there are new tests for the error handling and no old tests were modified at all
3) the Cython expert reviewer didn't express any strong objections overall
4) CI is passing/100 % patch diff
5) this contains a bug fix + regression test

That seems like a fair number of points in favor here. Any change in behavior from before would be untested behavior, which by some definitions is already broken anyway.

So, I'll merge it, thanks! 

If you could add a release note and perhaps open an issue to remove the test skips Ralf mentioned that might be helpful.

**Comment by tylerjereddy** [comment]
Thanks @andyfaff et al. -- note that I"ve also switched the title/label to clarify that this is a bug fix, though I suppose it is equally an enhancement for clarity/understanding.

In any case, feel free to re-label, etc. as you see appropriate.

**Comment by andyfaff** [comment]
I added a release note.

File: scipy/optimize/setup.py [hunk 7]
```diff
@@ -73,7 +73,7 @@ def configuration(parent_package='', top_path=None):
     ext._pre_build_hook = pre_build_hook
 
     sources = ['moduleTNC.c', 'tnc.c']
-    config.add_extension('_moduleTNC',
+    config.add_extension('moduleTNC',
```
Comment (rgommers) [inline scipy/optimize/setup.py:76]: @andyfaff I just ran into this - why did you remove the underscore? That seems like a bug to me, it's not a public submodule.

Comment (rgommers) [reply scipy/optimize/setup.py:76]: Opened gh-15270 to revert this change.
